client/db/src/lib.rs

```diff
@@ -1614,6 +1621,41 @@ impl<Block: BlockT> Backend<Block> {
 						children,
 					);
 				}
+
+				if let Some((mut start, end)) = block_gap {
+					if number == start {
+						start += One::one();
+						utils::insert_number_to_key_mapping(
+							&mut transaction,
+							columns::KEY_LOOKUP,
+							number,
+							hash,
+						)?;
+					}
+					if start > end {
+						transaction.remove(columns::META, meta_keys::BLOCK_GAP);
+						block_gap = None;
+						debug!(target: "db", "Removed block gap.");
+					} else {
+						block_gap = Some((start, end));
+						debug!(target: "db", "Update block gap. {:?}", block_gap);
+						transaction.set(
+							columns::META,
+							meta_keys::BLOCK_GAP,
+							&(start, end).encode(),
+						);
+					}
+				} else if number > best_num + One::one() &&
```

Comment:
> The point checked here is first block imported after a warp sync triggers gap download?

This is triggered when we import a warp block. I.e. we start at genesis and warp to block 10000. Block 10000 is imported with state. Gap is detected. No need to wait for the next block. 

With `number > best_num` it would be a false detection. E.g. block 43 is imported after 42 is the normal case. Only if we import 44 after 42 there's a gap.